core/src/data_sources/data_source.rs

```diff
@@ -730,25 +792,184 @@ impl DataSource {
             .buffer_unordered(16)
             .map(|r| match r {
-                Err(e) => Err(anyhow!("DataSource document retrieval error: {}", e))?,
+                Err(e) => Err(anyhow!("datasource document retrieval error: {}", e))?,
                 Ok(r) => r,
             })
             .try_collect::<Vec<_>>()
             .await?;
 
-        // Merge the chunks with the documents.
-        let mut documents = documents
-            .into_iter()
-            .map(|mut d| {
-                let chunks = chunks
-                    .iter()
-                    .filter(|(document_id, _)| document_id == &d.document_id)
-                    .map(|(_, c)| c.clone())
-                    .collect::<Vec<Chunk>>();
-                d.chunks = chunks;
-                d
-            })
-            .collect::<Vec<_>>();
+        let l_qdrant_client = Arc::new(Mutex::new(qdrant_client));
+        let mut documents = match target_document_tokens {
+            Some(target) => {
+                futures::stream::iter(documents)
+                    .map(|mut d| {
+                        let mut chunks = chunks
+                            .iter()
+                            .filter(|(document_id, _)| document_id == &d.document_id)
+                            .map(|(_, c)| c.clone())
+                            .collect::<Vec<Chunk>>();
+                        let qdrant_client = l_qdrant_client.clone();
+                        let collection = self.qdrant_collection();
+                        let chunk_size = self.config.max_chunk_size;
+                        tokio::spawn(async move {
+                            let qdrant_client = qdrant_client.lock().await;
+                            let mut offset_set = std::collections::HashSet::new();
+                            for chunk in chunks.iter() {
+                                offset_set.insert(chunk.offset);
+                            }
+                            let current_length = chunks.len() * chunk_size;
+                            if (target as i64 - current_length as i64) < 0 {
+                                d.chunks = chunks;
+                                return Ok(d);
+                            }
+                            let new_offsets = target_document_tokens_offsets(
+                                chunks.iter().map(|c| c.offset).collect(),
+                                target - current_length,
+                                chunk_size,
+                                d.chunk_count,
+                            );
+                            let offset_values: Vec<i64> = new_offsets
+                                .keys()
+                                .cloned()
+                                .collect::<Vec<usize>>()
+                                .into_iter()
+                                .map(|o| o as i64)
+                                .collect();
+                            let no_new_offsets = offset_values.len() as u32;
```

Comment:
new_offsets_count (out of consistency with the rest of the codebase where we use _count more than no_)